### ISSUE TITLE
SEGV in torch.sparse.mm with malformed CSR input and zero-column dense operand

### ISSUE THREAD

**Issue by yyds1233** [pytorch/pytorch#190045]
## Summary

`torch.sparse.mm` can crash on CPU with a malformed CSR matrix whose shape is `(3, 0)` but whose CSR metadata reports three stored values in column `0`. Multiplying it by a dense `(0, 2)` tensor enters the sparse CSR addmm path and triggers an ASan-detected null read in the CPU BLAS AXPY helper.

## Reproducer

Save the following as `poc.py` and run:

```bash
python3 poc.py
```

```python
import torch

crow = torch.tensor([0, 3, 3, 3], dtype=torch.int64)
col = torch.tensor([0, 0, 0], dtype=torch.int64)
values = torch.tensor([0.0, 0.0, 0.0], dtype=torch.float32)
mat1 = torch.sparse_csr_tensor(crow, col, values, size=(3, 0), check_invariants=False)
mat2 = torch.empty((0, 2), dtype=torch.float32)
torch.sparse.mm(mat1, mat2)
```

## ASAN-report：

```bash
AddressSanitizer:DEADLYSIGNAL
=================================================================
==1336637==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000000 (pc 0x7f8f11804cc0 bp 0x7ffe8a9ad770 sp 0x7ffe8a9ad690 T0)
==1336637==The signal is caused by a READ memory access.
==1336637==Hint: address points to the zero page.
    #0 0x7f8f11804cc0 in at::native::cpublas::(anonymous namespace)::cpublas_axpy_impl(c10::ScalarType, long, c10::Scalar const&, void const*, long, void*, long)::$_4::operator()() const::'lambda5'()::operator()() const /root/pytorch/aten/src/ATen/native/cpu/BlasKernel.cpp:529:5
    #1 0x7f8f11804cc0 in at::native::cpublas::(anonymous namespace)::cpublas_axpy_impl(c10::ScalarType, long, c10::Scalar const&, void const*, long, void*, long)::$_4::operator()() const /root/pytorch/aten/src/ATen/native/cpu/BlasKernel.cpp:529:5
    #2 0x7f8f11804cc0 in at::native::cpublas::(anonymous namespace)::cpublas_axpy_impl(c10::ScalarType, long, c10::Scalar const&, void const*, long, void*, long) /root/pytorch/aten/src/ATen/native/cpu/BlasKernel.cpp:529:5
    #3 0x7f8f06318c99 in void at::native::DispatchStub<void (*)(c10::ScalarType, long, c10::Scalar const&, void const*, long, void*, long), at::native::cpublas::axpy_stub_DECLARE_DISPATCH_type>::operator()<c10::ScalarType const&, long&, float&, float const*&, long&, float*&, long&>(c10::DeviceType, c10::ScalarType const&, long&, float&, float const*&, long&, float*&, long&) /root/pytorch/aten/src/ATen/native/DispatchStub.h:252:12
    #4 0x7f8f06318c99 in void at::native::cpublas::axpy<float>(long, float, float const*, long, float*, long) /root/pytorch/aten/src/ATen/native/CPUBlas.h:171:3
    #5 0x7f8f06318c99 in void at::native::addmm_out_sparse_csr_native_cpu<float>(at::Tensor const&, at::Tensor const&, at::Tensor const&, c10::Scalar, c10::Scalar)::'lambda'()::operator()() const::'lambda0'()::operator()() const::'lambda'(long, long)::operator()(long, long) const /root/pytorch/aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp:549:3
    #6 0x7f8f063174dc in void at::parallel_for<void at::native::addmm_out_sparse_csr_native_cpu<float>(at::Tensor const&, at::Tensor const&, at::Tensor const&, c10::Scalar, c10::Scalar)::'lambda'()::operator()() const::'lambda0'()::operator()() const::'lambda'(long, long)>(long, long, long, void at::native::addmm_out_sparse_csr_native_cpu<float>(at::Tensor const&, at::Tensor const&, at::Tensor const&, c10::Scalar, c10::Scalar)::'lambda'()::operator()() const::'lambda0'()::operator()() const::'lambda'(long, long) const&) /root/pytorch/aten/src/ATen/Parallel-inl.h:29:5
    #7 0x7f8f063174dc in void at::native::addmm_out_sparse_csr_native_cpu<float>(at::Tensor const&, at::Tensor const&, at::Tensor const&, c10::Scalar, c10::Scalar)::'lambda'()::operator()() const::'lambda0'()::operator()() const /root/pytorch/aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp:549:3
    #8 0x7f8f063174dc in void at::native::addmm_out_sparse_csr_native_cpu<float>(at::Tensor const&, at::Tensor const&, at::Tensor const&, c10::Scalar, c10::Scalar)::'lambda'()::operator()() const /root/pytorch/aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp:549:3
    #9 0x7f8f063174dc in void at::native::addmm_out_sparse_csr_native_cpu<float>(at::Tensor const&, at::Tensor const&, at::Tensor const&, c10::Scalar, c10::Scalar) /root/pytorch/aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp:549:3
    #10 0x7f8f0630f1ce in at::native::addmm_out_sparse_compressed_cpu(at::Tensor const&, at::Tensor const&, at::Tensor const&, c10::Scalar const&, c10::Scalar const&, at::Tensor&)::$_10::operator()() const::'lambda0'()::operator()() const /root/pytorch/aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp:704:9
    #11 0x7f8f0630f1ce in at::native::addmm_out_sparse_compressed_cpu(at::Tensor const&, at::Tensor const&, at::Tensor const&, c10::Scalar const&, c10::Scalar const&, at::Tensor&)::$_10::operator()() const /root/pytorch/aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp:704:9
    #12 0x7f8f0630f1ce in at::native::addmm_out_sparse_compressed_cpu(at::Tensor const&, at::Tensor const&, at::Tensor const&, c10::Scalar const&, c10::Scalar const&, at::Tensor&) /root/pytorch/aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp:704:9
    #13 0x7f8f06a7fff9 in at::Tensor& c10::callUnboxedKernelFunction<at::Tensor&, at::Tensor const&, at::Tensor const&, at::Tensor const&, c10::Scalar const&, c10::Scalar const&, at::Tensor&>(void*, c10::OperatorKernel*, c10::DispatchKeySet, at::Tensor const&, at::Tensor const&, at::Tensor const&, c10::Scalar const&, c10::Scalar const&, at::Tensor&) /root/pytorch/aten/src/ATen/core/boxing/KernelFunction_impl.h:104:10
    #14 0x7f8f06a7fff9 in at::Tensor& c10::KernelFunction::call<at::Tensor&, at::Tensor const&, at::Tensor const&, at::Tensor const&, c10::Scalar const&, c10::Scalar const&, at::Tensor&>(c10::OperatorHandle const&, c10::DispatchKeySet, at::Tensor const&, at::Tensor const&, at::Tensor const&, c10::Scalar const&, c10::Scalar const&, at::Tensor&) const /root/pytorch/aten/src/ATen/core/boxing/KernelFunction_impl.h:173:14
    #15 0x7f8f06a7fff9 in at::Tensor& c10::Dispatcher::call<at::Tensor&, at::Tensor const&, at::Tensor const&, at::Tensor const&, c10::Scalar const&, c10::Scalar const&, at::Tensor&>(c10::TypedOperatorHandle<at::Tensor& (at::Tensor const&, at::Tensor const&, at::Tensor const&, c10::Scalar const&, c10::Scalar const&, at::Tensor&)> const&, at::Tensor const&, at::Tensor const&, at::Tensor const&, c10::Scalar const&, c10::Scalar const&, at::Tensor&) const /root/pytorch/aten/src/ATen/core/dispatch/Dispatcher.h:827:26
    #16 0x7f8f06a7fff9 in c10::TypedOperatorHandle<at::Tensor& (at::Tensor const&, at::Tensor const&, at::Tensor const&, c10::Scalar const&, c10::Scalar const&, at::Tensor&)>::call(at::Tensor const&, at::Tensor const&, at::Tensor const&, c10::Scalar const&, c10::Scalar const&, at::Tensor&) const /root/pytorch/aten/src/ATen/core/dispatch/Dispatcher.h:615:41
    #17 0x7f8f06a7fff9 in at::_ops::addmm_out::call(at::Tensor const&, at::Tensor const&, at::Tensor const&, c10::Scalar const&, c10::Scalar const&, at::Tensor&) /root/pytorch/build/aten/src/ATen/Operators_0.cpp:6219:15
    #18 0x7f8f0631ecd9 in at::addmm_out(at::Tensor&, at::Tensor const&, at::Tensor const&, at::Tensor const&, c10::Scalar const&, c10::Scalar const&) /root/pytorch/build/aten/src/ATen/ops/addmm.h:28:12
    #19 0x7f8f0631ecd9 in at::native::addmm_sparse_compressed_dense(at::Tensor const&, at::Tensor const&, at::Tensor const&, c10::Scalar const&, c10::Scalar const&) /root/pytorch/aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp:746:3
AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV /root/pytorch/aten/src/ATen/native/cpu/BlasKernel.cpp:529:5 in at::native::cpublas::(anonymous namespace)::cpublas_axpy_impl(c10::ScalarType, long, c10::Scalar const&, void const*, long, void*, long)::$_4::operator()() const::'lambda5'()::operator()() const
==1336637==ABORTING
```

## Expected behavior

Invalid Python-level inputs should be rejected deterministically with a Python exception before entering native kernels. They should not lead to an ASan-detected FPE, segmentation fault, or out-of-bounds native memory access.

## Root cause analysis

The CSR tensor is constructible when invariant checking is disabled: it has zero columns but contains `col_indices=[0,0,0]` and `values=[0,0,0]`. `torch.sparse.mm` later trusts this malformed structure and enters the sparse CSR matrix multiplication path. The ASan stack shows a read SEGV in `cpublas_axpy_impl`, reached from `addmm_out_sparse_csr_native_cpu`, indicating that the malformed compressed structure is not rejected before it is consumed by the multiplication kernel.

### Versions

PyTorch version: 2.10.0a0+gite5a766e
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04.2) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 4.1.2
Libc version: glibc-2.35

Python version: 3.10.12 (main, Aug 15 2025, 14:32:43) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.4.0-200-generic-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: 
Nvidia driver version: Could not collect
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      48 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             128
On-line CPU(s) list:                0-127
Vendor ID:                          AuthenticAMD
Model name:                         AMD Ryzen Threadripper PRO 5995WX 64-Cores
CPU family:                         25
Model:                              8
Thread(s) per core:                 2
Core(s) per socket:                 64
Socket(s):                          1
Stepping:                           2
Frequency boost:                    enabled
CPU max MHz:                        2700.0000
CPU min MHz:                        1800.0000
BogoMIPS:                           5390.20
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 invpcid_single hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr wbnoinvd arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca
Virtualization:                     AMD-V
L1d cache:                          2 MiB (64 instances)
L1i cache:                          2 MiB (64 instances)
L2 cache:                           32 MiB (64 instances)
L3 cache:                           256 MiB (8 instances)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-127
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Retpolines; IBPB conditional; IBRS_FW; STIBP always-on; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] numpy==2.2.6
[pip3] nvidia-cublas-cu12==12.8.4.1
[pip3] nvidia-cuda-cupti-cu12==12.8.90
[pip3] nvidia-cuda-nvrtc-cu12==12.8.93
[pip3] nvidia-cuda-runtime-cu12==12.8.90
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cufft-cu12==11.3.3.83
[pip3] nvidia-curand-cu12==10.3.9.90
[pip3] nvidia-cusolver-cu12==11.7.3.90
[pip3] nvidia-cusparse-cu12==12.5.8.93
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-nccl-cu12==2.27.3
[pip3] nvidia-nvjitlink-cu12==12.8.93
[pip3] nvidia-nvtx-cu12==12.8.90
[pip3] optree==0.18.0
[pip3] torch==2.10.0a0+gite5a766e
[pip3] torchaudio==2.10.0a0
[pip3] torchcodec==0.9.0a0
[pip3] triton==3.4.0
[conda] Could not collect

cc @nikitaved @pearu @cpuhrsch @amjames @bhosmer @jcaip

**Comment by CodersAcademy006** [comment]
The crash is a missing bound on the column index in the CPU CSR mm path (`addmm_out_sparse_csr_native_cpu`, `aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp`, ~L569-581). For each stored value the inner loop does `col = col_indices_accessor[i]` and then `axpy(dim_k, ..., dense_ptr + col * dense_stride0, ...)` without ever checking `col` against the contraction dimension. In this reproducer `mat1` is a `(3, 0)` CSR built with `check_invariants=False` whose `crow=[0,3,3,3]` claims three stored values in column 0, while `mat2` is `(0, 2)`, so the contraction dim is 0 and `dense_ptr` backs an empty tensor. `col=0` then makes `axpy` read `dim_k` elements starting at `dense_ptr`, i.e. from empty storage, which is the zero-page read in the ASan report. Since it is reachable from Python (even if via `check_invariants=False`), an early return when the contraction dim `mat2.size(0)` is 0, or a bound check that `col < mat2.size(0)`, would convert this into a proper error rather than a SEGV. @cpuhrsch

**Comment by Arindamsamanta2004** [comment]
I looked into this, the CPU path in addmm_out_sparse_compressed_cpu is missing a guard for zero contraction dimension. The CUDA side already handles this (PR #66485 added an early return in addmm_out_sparse_csr_dense_cuda), but the CPU path never got the same treatment.
As @CodersAcademy006 suggested, an early return when mat1.size(1) == 0 right after the existing result.numel() == 0 check fixes this. I verified this locally, reproducer returns correct (3, 2) zeros instead of segfaulting and existing tests pass clean.
Happy to raise a PR for this if no one's already working on it. cc @cpuhrsch @amjames @CodersAcademy006

**Comment by amjames** [comment]
This is similar to #190045 and essentially has the same resolution. If the invalid read can only be triggered by a crafted input that does not pass invariant checks then I don't really consider it a real bug. 

If the reproducer performs a zero-page read on the dense tensor's buffer with a valid CSR input eg 

```python
import torch


mat1 = torch.empty((3, 0)).to_sparse_csr()
mat2 = torch.empty((0, 2), dtype=torch.float32)
torch.sparse.mm(mat1, mat2)
```

Then I would consider it a real problem.

@Arindamsamanta2004 #66485 does not introduce an early return for length zero reduction dim, it is an early exit on zero sized output. A check on the reduction dim is redundant as we already check that the matrix sizes are valid for a matmul, and we drive the computation by the sparse elements which are addressable. If the input in the original reproducer was not malformed it would either have shape (3, 3) and fail the matmul shape checks, or it would have nnz=0 and the loop would terminate immediately.

**Comment by amjames** [comment]
Closing as not planned (see above)
